test/UnitTest/Microsoft.AspNetCore.OData.Test/Routing/ODataActionSelectorTest.cs

```diff
@@ -0,0 +1,248 @@
+            // Arrange
+            ODataActionSelectorTestHelper.SetupActionSelector(expectedAction.ControllerType, expectedAction.ActionName,
+                out var routeBuilder, out var actionSelector, out var actionDescriptors);
+            var routeContext = ODataActionSelectorTestHelper.SetupRouteContext(routeBuilder, expectedAction.ActionName, routeDataValues,
+                requestData.Method, requestData.Body);
+
+            // Act
+            var action = actionSelector.SelectBestCandidate(routeContext, actionDescriptors);
+
+            // Assert
+            var method = expectedAction.ControllerType.GetMethod(expectedAction.ActionName, expectedAction.Signature);
+            Assert.NotNull(action);
+            Assert.True(ODataActionSelectorTestHelper.ActionMatchesMethod(action, method));
+        }
+
+        [Theory]
+        [MemberData(nameof(ScenariosWithNoCorrectAction))]
+        public void SelectBestCandidate_WhenNoActionWithMatchingParameters_ReturnsNull(
+            Dictionary<string, object> routeDataValues,
+            (string Method, string Body) requestData,
+            (Type ControllerType, string ActionName) actionData)
+        {
+            // Arrange
+            ODataActionSelectorTestHelper.SetupActionSelector(actionData.ControllerType, actionData.ActionName, out var routeBuilder,
+                out var actionSelector, out var actionDescriptors);
+            var routeContext = ODataActionSelectorTestHelper.SetupRouteContext(routeBuilder, actionData.ActionName, routeDataValues,
+                requestData.Method, requestData.Body);
+
+            // Act
+            var action = actionSelector.SelectBestCandidate(routeContext, actionDescriptors);
+
+            // Assert
+            Assert.True(action == null);
+        }
+    }
+
+    public class SingleKeyController : TestODataController
+    {
+        public void Get(int key) { }
+        public void Get() { }
+    }
+
+    public class PathAndQueryController : TestODataController
+    {
+        public void Get(ODataPath path, ODataQueryOptions queryOptions) { }
+
+        public void Get(ODataPath path) { }
+    }
+
+    public class KeyAndQueryController : TestODataController
+    {
+        public void Get(int key, ODataQueryOptions queryOptions) { }
+
+        public void Get(int key) { }
+    }
+
+    public class KeyAndRelatedKeyController : TestODataController
+    {
+        public void Get(int key, int relatedKey) { }
+
+        public void Get(int key) { }
+    }
+
+    public class KeyAndRelatedKeyAndPathController : TestODataController
+    {
+        public void Get(int key, int relatedKey, ODataPath path) { }
+
+        public void Get(int key, int relatedKey) { }
+
+        public void Get(int key, int relatedKey, int navigationProperty) { }
+
+        public void Get(int key, ODataPath path) { }
+    }
+
+    public class KeyBodyController : TestODataController
+    {
+        public void Put(int key, Delta<object> dt) { }
```

Comment:
@habbes Should this be patch?

Reply:
It was meant to be `Put`. But I assume you're asking that because I used `Delta<object>`. I can change it to `Patch` to match the standard practice, it shouldn't affect the validity of the test.